src/bootstrap/dist.rs

```diff
@@ -1897,6 +1899,26 @@ fn add_env(builder: &Builder<'_>, cmd: &mut Command, target: TargetSelection) {
     }
 }
 
+fn install_llvm_file(builder: &Builder<'_>, source: &Path, destination: &Path) {
+    if source.as_os_str().is_empty() {
+        return;
+    }
+
+    // After LLVM is built, we modify (instrument or optimize) the libLLVM.so library file.
+    // This is not done in-place so that the built LLVM files are not "tainted" with BOLT.
+    // We perform the instrumentation/optimization here, on the fly, just before they are being
+    // packaged into some destination directory.
+    let postprocessed = if builder.config.llvm_bolt_profile_generate {
+        builder.ensure(BoltInstrument::new(source.to_path_buf()))
+    } else if let Some(path) = &builder.config.llvm_bolt_profile_use {
+        builder.ensure(BoltOptimize::new(source.to_path_buf(), path.into()))
```

Comment:
Just to double check, this is not going to rerun bolt multiple times for multiple calls to install_llvm_file, right? (I think we call it multiple times to copy it to different dirs.)

It might be interesting to get metrics for the BoltInstrument/BoltOptimize steps as well.

Reply:
> It might be interesting to get metrics for the BoltInstrument/BoltOptimize steps as well.

Ah, nevermind. I see they are already printed, just not as part of the stage-build summary.

Reply:
That's correct, it won't run it multiple times, it is cached. It can be seen in https://github.com/rust-lang-ci/rust/actions/runs/4184866597/jobs/7251109192 - BOLT is only run twice (you can grep for `executable detected`). Before caching was implemented, it was executed multiple times, as it is indeed copied multiple times (stage1/stage2).

Good idea with the metrics, but I don't think that we need to put it into the final table, it's already quite long. The CI log now prints all metrics of all bootstrap invocations, so we can just look into it when we need to:

<details>
<summary>Metrics log</summary>

```
2023-02-15T16:16:37.8371756Z root INFO: Build step durations
2023-02-15T16:16:37.8372133Z root                                                              2452.46s
2023-02-15T16:16:37.8375395Z .bootstrap::compile::Assemble                                        0.08s
2023-02-15T16:16:37.8375850Z ..bootstrap::compile::Assemble                                       0.05s
2023-02-15T16:16:37.8376305Z ...bootstrap::compile::Assemble                                      0.00s
2023-02-15T16:16:37.8376725Z ...bootstrap::compile::Rustc                                         0.04s
2023-02-15T16:16:37.8377157Z ....bootstrap::compile::Std                                          0.03s
2023-02-15T16:16:37.8377588Z .....bootstrap::compile::StartupObjects                              0.00s
2023-02-15T16:16:37.8378031Z .....bootstrap::builder::Builder::sysroot_libdir::Libdir             0.00s
2023-02-15T16:16:37.8378471Z ......bootstrap::compile::Sysroot                                    0.00s
2023-02-15T16:16:37.8378880Z .....bootstrap::compile::StdLink                                     0.00s
2023-02-15T16:16:37.8395973Z ....bootstrap::native::Llvm                                          0.00s
2023-02-15T16:16:37.8396421Z ....bootstrap::compile::RustcLink                                    0.00s
2023-02-15T16:16:37.8396859Z ...bootstrap::native::Lld                                            0.00s
2023-02-15T16:16:37.8397269Z ...bootstrap::compile::Sysroot                                       0.00s
2023-02-15T16:16:37.8397696Z ...bootstrap::builder::Builder::sysroot_libdir::Libdir               0.00s
2023-02-15T16:16:37.8398147Z ...bootstrap::tool::LldWrapper                                       0.00s
2023-02-15T16:16:37.8398565Z ....bootstrap::tool::ToolBuild                                       0.00s
2023-02-15T16:16:37.8398966Z ..bootstrap::compile::Rustc                                          0.01s
2023-02-15T16:16:37.8399372Z ...bootstrap::compile::Std                                           0.00s
2023-02-15T16:16:37.8399797Z ....bootstrap::compile::StartupObjects                               0.00s
2023-02-15T16:16:37.8400213Z ....bootstrap::native::Sanitizers                                    0.00s
2023-02-15T16:16:37.8400684Z ....bootstrap::compile::StdLink                                      0.00s
2023-02-15T16:16:37.8401100Z ...bootstrap::compile::RustcLink                                     0.00s
2023-02-15T16:16:37.8401503Z ..bootstrap::compile::Sysroot                                        0.03s
2023-02-15T16:16:37.8401936Z ..bootstrap::builder::Builder::sysroot_libdir::Libdir                0.00s
2023-02-15T16:16:37.8402366Z ..bootstrap::tool::LldWrapper                                        0.00s
2023-02-15T16:16:37.8407994Z ...bootstrap::tool::ToolBuild                                        0.00s
2023-02-15T16:16:37.8408421Z .bootstrap::compile::Std                                             0.00s
2023-02-15T16:16:37.8408838Z ..bootstrap::compile::StartupObjects                                 0.00s
2023-02-15T16:16:37.8409558Z ..bootstrap::compile::StdLink                                        0.00s
2023-02-15T16:16:37.8415753Z .bootstrap::compile::Assemble                                     2452.27s
2023-02-15T16:16:37.8416217Z ..bootstrap::compile::Assemble                                    1392.92s
2023-02-15T16:16:37.8416674Z ...bootstrap::compile::Assemble                                      0.00s
2023-02-15T16:16:37.8417092Z ...bootstrap::compile::Rustc                                      1116.02s
2023-02-15T16:16:37.8417489Z ....bootstrap::compile::Std                                          0.25s
2023-02-15T16:16:37.8418136Z .....bootstrap::compile::StartupObjects                              0.00s
2023-02-15T16:16:37.8418685Z .....bootstrap::builder::Builder::sysroot_libdir::Libdir             0.00s
2023-02-15T16:16:37.8419123Z ......bootstrap::compile::Sysroot                                    0.00s
2023-02-15T16:16:37.8419544Z .....bootstrap::compile::StdLink                                     0.00s
2023-02-15T16:16:37.8419950Z ....bootstrap::native::Llvm                                       1097.91s
2023-02-15T16:16:37.8420451Z ....bootstrap::compile::RustcLink                                    0.01s
2023-02-15T16:16:37.8421080Z ...bootstrap::native::Lld                                           16.41s
2023-02-15T16:16:37.8421469Z ...bootstrap::compile::Sysroot                                       0.00s
2023-02-15T16:16:37.8421995Z ...bootstrap::builder::Builder::sysroot_libdir::Libdir               0.00s
2023-02-15T16:16:37.8422403Z ...bootstrap::tool::LldWrapper                                       0.23s
2023-02-15T16:16:37.8422789Z ....bootstrap::tool::ToolBuild                                       0.23s
2023-02-15T16:16:37.8423169Z ...bootstrap::dist::BoltInstrument                                 245.12s
2023-02-15T16:16:37.8423559Z ..bootstrap::compile::Rustc                                       1036.60s
2023-02-15T16:16:37.8423934Z ...bootstrap::compile::Std                                         137.11s
2023-02-15T16:16:37.8424402Z ....bootstrap::compile::StartupObjects                               0.00s
2023-02-15T16:16:37.8424778Z ....bootstrap::native::Sanitizers                                    0.00s
2023-02-15T16:16:37.8425137Z ....bootstrap::compile::StdLink                                      0.00s
2023-02-15T16:16:37.8425490Z ...bootstrap::compile::RustcLink                                     0.01s
2023-02-15T16:16:37.8425864Z ..bootstrap::compile::Sysroot                                        0.00s
2023-02-15T16:16:37.8426233Z ..bootstrap::builder::Builder::sysroot_libdir::Libdir                0.00s
2023-02-15T16:16:37.8426608Z ..bootstrap::tool::LldWrapper                                        7.41s
2023-02-15T16:16:37.8426983Z ...bootstrap::tool::ToolBuild                                        7.41s
2023-02-15T16:16:37.8427708Z .bootstrap::compile::Std                                             0.00s
2023-02-15T16:16:37.8428117Z ..bootstrap::compile::StartupObjects                                 0.00s
2023-02-15T16:16:37.8428561Z ..bootstrap::compile::StdLink                                        0.00s
```
</details>

BOLT instrumentation takes < 5 minutes, optimization takes < 1 minute.